lib/json/json.go

```diff
@@ -496,18 +501,22 @@ func decode(thread *starlark.Thread, b *starlark.Builtin, args starlark.Tuple, k
 		x := recover()
 		switch x := x.(type) {
 		case failure:
-			err = fmt.Errorf("json.decode: at offset %d, %s", i, x)
+			if d != nil {
```

Comment:
One thing to be wary of in this sentinel-based error reporting is the potential for inadvertently creating an exception handler in the language. Imagine a variant of decode that under some circumstance called a user-supplied function. By calling `decode(fn, default=123)`, the user could call fn, handle any error that it raises, and replace it with a sentinel value. This could potentially be used to continue after a function that leaves the thread in a bad state, or to continue and perhaps retry after some error (such as a one-shot equivalent of the limit on computation steps).

I don't see any problem with this logic as it stands, but if we use this approach more widely we will need to think carefully about each case.

Reply:
Acknowledged - but in this case, we are matching that the failure type is `failure` which is declared locally and not exported, so any other source of panic would fall through to

```
		default:
			panic(x) // unexpected panic
```

I will add a comment to make this more clear.